### ISSUE TITLE
get event loop for the fastapi

### ISSUE THREAD

**Issue by timofei-durakov** [tiangolo/fastapi#825]
Is there any way to get event or provide event loop for the fast api app. I would like to run low level tcp server(asyncio low-level protocols) and fast api within the same process. So I need to reuse the same event loop for both of them. Is it possible doing somehow?

**Comment by prostomarkeloff** [comment]
Yes, i think it's possible. You can start the fastapi application by hypercorn.
Source: https://pgjones.gitlab.io/hypercorn/api_usage.html 

Your code should look like that:
```python
# some definitons

loop = asyncio.get_event_loop()

loop.create_task(serve(app, config))  # run fastapi
loop.create_task(my_super_app()) # run your app
loop.run_forever()  # start event loop
```

I don't know how to do something like that with `uvicorn`.

**Comment by dmontagu** [comment]
Unfortunately it looks like uvicorn doesn't expose a great way to specify the event loop to use. However, in my experience uvicorn is *substantially* faster than hypercorn, so may be worth a little extra pain to make things work. Especially if your TCP server is going to be interacting with the python event loop with any substantial frequency (which would not surprise me).

You can get the event loop being used by FastAPI by calling `asyncio.get_running_loop()` in a function that will be called with that event loop; e.g., inside an `async def` [server startup event](https://fastapi.tiangolo.com/tutorial/events/).

If possible, my recommendation would be to put the startup of your tcp server into a FastAPI startup event so it can run in the same uvicorn loop. This would probably be the easiest way to get things to work.

If you need to start the TCP server prior to the FastAPI server for some reason, it will probably be more complicated, but likely not impossible.

**Comment by euri10** [comment]
You can pass your loop into uvicorn's Config

**Comment by dmontagu** [comment]
@euri10 A cursory look at `config.py` made me think this would cause an issue:

https://github.com/encode/uvicorn/blob/abd68704ea75cdbac8ffb1b5c59578a17a716a56/uvicorn/config.py#L319

In particular, that function is called by `uvicorn.run`, and it does a `dict.__getitem__` with the provided `loop` value. But the relevant dict only has string keys. So I don't think you are supposed to pass a loop *instance* to the uvicorn config.

Is there a different way to make this work?

**Comment by euri10** [comment]
https://github.com/encode/uvicorn/pull/510 the last test uses the pattern I'm thinking of

**Comment by dmontagu** [comment]
Cool, thanks for looking into that @euri10.

@timofei-durakov if you want to configure the event loop yourself, based on the example @euri10 provided it should look like:
```python
config = Config(app=fastapi_app, loop=loop_instance)
server = Server(config=config)
server.run()
```
I haven't verified that will work though.

That said, if you don't want to figure out precisely how to manually set up the event loop for maximum performance, it may still be easier (if possible) to set up the TCP server from inside a startup event of the FastAPI server.

**Comment by tiangolo** [comment]
Thanks for all the help here everyone! :bowing_man: :cake: 

@timofei-durakov does that solve your issue?

**Comment by lsaint** [comment]
@dmontagu 

> If possible, my recommendation would be to put the startup of your tcp server into a FastAPI startup event so it can run in the same uvicorn loop. This would probably be the easiest way to get things to work.


let's say, i have a new event loop,  to run a function every seconds, how to add it in this way?
